### PR TITLE
Np 2399 use apikey

### PR DESCRIPTION
**Pull request by svenngjeto**

### CODE REVIEW

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 1]
```diff
@@ -137,4 +168,42 @@ private URI addPathSegments(List<String> pathSegments) throws URISyntaxException
         }
         return URLEncodedUtils.parsePathSegments(path);
     }
+
+
+    private Optional<String> getCrossRefApiPlusToken() {
```
Comment (axthosarouris) [inline doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java:173]: We have a SecretsReader in nva commons

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 2]
```diff
@@ -137,4 +168,42 @@ private URI addPathSegments(List<String> pathSegments) throws URISyntaxException
...
+
+        Optional<String> secret = Optional.empty();
```
Comment (axthosarouris) [inline doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java:175]: We have a SecretsReader in nva-commons

File: template.yaml [hunk 3]
```diff
@@ -116,19 +121,53 @@ Resources:
...
+                  - logs:PutLogEvents
+                Resource: !Sub 'arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:*:*'
+        - PolicyName: readSecret
+          PolicyDocument:
+            Version: 2012-10-17
+            Statement:
+              - Effect: Allow
+                Action:
+                  - secretsmanager:GetSecretValue
+                Resource: !Sub 'arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:*'
```
Comment (axthosarouris) [inline template.yaml:153]: better if we allow access to one specific secret

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 4]
```diff
@@ -81,15 +98,29 @@ public CrossRefClient(HttpClient httpClient) {
...
-            .GET()
-            .build();
+                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
+                .header(HttpHeaders.USER_AGENT, CROSSREF_USER_AGENT)
+                .timeout(Duration.ofSeconds(TIMEOUT_DURATION))
+                .GET()
+                .build();
     }
 
+//    private HttpRequest createRequest(URI doiUri) {
```
Comment (axthosarouris) [inline doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java:108]: remove commented out code

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 5]
```diff
@@ -137,4 +169,49 @@ private URI addPathSegments(List<String> pathSegments) throws URISyntaxException
...
+    private Optional<String> getCrossRefApiPlusToken() {
+
+        try {
+            return Optional.ofNullable(new  SecretsReader().fetchSecret(secretName.get(), secretKey.get()));
+        } catch (ErrorReadingSecretException e) {
+            logger.error(EXCEPTION_READING_API_SECRET_FROM_AWS_SECRETS_MANAGER);
+            return Optional.empty();
+        }
+
+//        Optional<String> secret = Optional.empty();
```
Comment (brinxmat) [inline doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java:183]: Commented?

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 2 above)
Comment (svenngjeto) [reply doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java]: Using SecretsReader

File: template.yaml (hunk 3 above)
Comment (svenngjeto) [reply template.yaml]: Limited

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 5 above)
Comment (svenngjeto) [reply doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java]: Gone

**Review by brinxmat — DISMISSED** [review]

File: doi-proxy/build.gradle [hunk 6]
```diff
@@ -3,4 +3,9 @@ dependencies {
 
     runtimeOnly group: 'org.glassfish.jersey.core', name: 'jersey-common', version: project.ext.jerseyVersion
     implementation group: 'jakarta.ws.rs', name: 'jakarta.ws.rs-api', version: project.ext.jakartaVersion
+
+    implementation group: 'com.amazonaws', name: 'aws-java-sdk-secretsmanager', version: project.ext.awsSdkVersion
+    implementation group: 'com.amazonaws', name: 'aws-java-sdk-sts', version: project.ext.awsSdkVersion
```
Comment (axthosarouris) [inline doi-proxy/build.gradle:8]: Where are we using sts?

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 7]
```diff
@@ -31,23 +36,35 @@
...
-
-    private static final String DOI_EXAMPLES = "10.1000/182, https://doi.org/10.1000/182";
-    public static final String ILLEGAL_DOI_MESSAGE = "Illegal DOI:%s. Valid examples:" + DOI_EXAMPLES;
     public static final String FETCH_ERROR = "CrossRefClient failed while trying to fetch:";
     public static final String CROSSREF_USER_AGENT =
             "nva-fetch-doi/1.0 (https://github.com/BIBSYSDEV/nva-fetch-doi; mailto:support@unit.no)";
-
-    private final transient HttpClient httpClient;
+    public static final String ADDING_TOKEN_IN_HEADER =
+            "CrossRefApiPlusToken.isPresent() == true, adding token in header";
```
Comment (axthosarouris) [inline doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java:43]: CrossRer Api token is present

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: doi-proxy/build.gradle (hunk 6 above)
Comment (svenngjeto) [reply doi-proxy/build.gradle]: we don't (anymore) is of couse replaced by nva-core/secrets

File: doi-proxy/build.gradle [hunk 8]
```diff
@@ -3,4 +3,9 @@ dependencies {
 
     runtimeOnly group: 'org.glassfish.jersey.core', name: 'jersey-common', version: project.ext.jerseyVersion
     implementation group: 'jakarta.ws.rs', name: 'jakarta.ws.rs-api', version: project.ext.jakartaVersion
+
+    implementation group: 'com.amazonaws', name: 'aws-java-sdk-secretsmanager', version: project.ext.awsSdkVersion
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply doi-proxy/build.gradle]: we don't (anymore) is of couse replaced by nva-core/secrets

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 7 above)
Comment (svenngjeto) [reply doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java]: Correct to somthing that looks like a sentence...

File: doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 9]
```diff
@@ -80,16 +97,21 @@ public CrossRefClient(HttpClient httpClient) {
...
-            .build();
+        HttpRequest.Builder builder = HttpRequest.newBuilder(doiUri)
+                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
+                .header(HttpHeaders.USER_AGENT, CROSSREF_USER_AGENT)
+                .timeout(Duration.ofSeconds(TIMEOUT_DURATION))
+                .GET();
+        if (getCrossRefApiPlusToken().isPresent()) {
+            logger.info(ADDING_TOKEN_IN_HEADER);
+            builder.setHeader(CROSSREF_PLUSAPI_HEADER,
+                    String.format(CROSSREF_PLUSAPI_AUTHORZATION_HEADER_BASE, getCrossRefApiPlusToken().get()));
```
Comment (axthosarouris) [inline doi-proxy/src/main/java/no/unit/nva/doi/CrossRefClient.java:108]: get() is not advised any more. use orElseThrow() or 

```
private HttpRequest createRequest(URI doiUri) {
        HttpRequest.Builder builder = HttpRequest.newBuilder(doiUri)
                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
                .header(HttpHeaders.USER_AGENT, CROSSREF_USER_AGENT)
                .timeout(Duration.ofSeconds(TIMEOUT_DURATION))
                .GET();
        getCrossRefApiPlusToken().ifPresent(crossRefToken-> addCrossRefTokenHeader(builder, crossRefToken));
        return builder.build();
    }

    private void addCrossRefTokenHeader(HttpRequest.Builder builder, String crossRefToken) {
        logger.info(ADDING_TOKEN_IN_HEADER);
        builder.header(CROSSREF_PLUSAPI_HEADER,
                       String.format(CROSSREF_PLUSAPI_AUTHORZATION_HEADER_BASE, crossRefToken));
    }
```

**Review by axthosarouris — APPROVED** [review]

**Review by janbj — APPROVED** [review]

**Pull request merged**